### PR TITLE
Radiation heat transfer

### PR DESCRIPTION
**Pull request by shahabgol**
# Description of the problem

- This PR modifies the convection heat transfer boundary condition to convection-radiation. This new BC is a combination of convection and radiation. If the Stefan-Boltzmann constant (with default value = 0) is set to 0, only the convection component is considered, whereas if the convection coefficient, h (with default value = 0), is set to 0, only the radiation component is considered. Otherwise, both the convection and radiation are significant on the boundary.

- It also adds "calculate temperature range" option to post-processing. By enabling "calculate temperature range", the solver reports the minimum and maximum temperature in the simulation domain.

# How Has This Been Tested?

- heat_transfer_laser application test is improved to test this new BC.

# Documentation

- boundary_conditions_multiphysics is updated.

### CODE REVIEW

File: include/solvers/postprocessing_cfd.h [hunk 1]
```diff
@@ -273,6 +273,23 @@ calculate_flow_rate(const DoFHandler<dim> &    dof_handler,
                     const Mapping<dim> &       mapping);
 
 
+/**
+ * @brief Calculates the minimum and maximum temperature in the simulation domain
+ *
+ * @param dof_handler. The argument used to get finite elements
```
Comment (blaisb) [inline include/solvers/postprocessing_cfd.h:279]:
```suggestion
 * @param dof_handler. The dof handler associated with the heat transfer problem.
```

File: include/solvers/postprocessing_cfd.h [hunk 2]
```diff
@@ -273,6 +273,23 @@ calculate_flow_rate(const DoFHandler<dim> &    dof_handler,
...
+ *
+ * @param face_quadrature_formula The face quadrature formula for the calculation
```
Comment (blaisb) [inline include/solvers/postprocessing_cfd.h:281]: This is not a face quadrature right? please review these comments. I think there is some copy paste errors there :)!

File: source/solvers/postprocessing_cfd.cc [hunk 3]
```diff
@@ -1040,3 +1040,81 @@ calculate_flow_rate(const DoFHandler<3> &                     dof_handler,
...
+  const FESystem<dim, dim> fe = dof_handler_ht.get_fe();
+
+  const unsigned int  n_q_points = quadrature_formula.size();
+  std::vector<double> local_temperature_values(n_q_points);
+
+  FEValues<dim> fe_values(mapping,
+                          fe,
+                          quadrature_formula,
+                          update_values | update_quadrature_points |
+                            update_JxW_values);
```
Comment (blaisb) [inline source/solvers/postprocessing_cfd.cc:1060]: Why update the quadrature points and the JxW values, you only use the values anyway? Might as well. update as minimally as possible.

**Review by blaisb — COMMENTED** [review]
Very good work. Do you plan on adding the capability to output this to a text file instead of just into the log? Or is that planned for a follow-up PR?

File: include/solvers/postprocessing_cfd.h (hunk 2 above)
Comment (shahabgol) [reply include/solvers/postprocessing_cfd.h:281]: True :) corrected.

**Comment by shahabgol** [comment]
> Very good work. Do you plan on adding the capability to output this to a text file instead of just into the log? Or is that planned for a follow-up PR?

Is it necessary to write minimum and maximum temperatures in a text file? I think we will only need it for testing. The rest of comments have been addressed.

**Comment by blaisb** [comment]
> > Very good work. Do you plan on adding the capability to output this to a text file instead of just into the log? Or is that planned for a follow-up PR?
> 
> Is it necessary to write minimum and maximum temperatures in a text file? I think we will only need it for testing. The rest of comments have been addressed.

Not necessary at all, you are right

**Comment by blaisb** [comment]
Could you add a quick unit test with an analytical solution for pure radiation BC (like a couette with a radiation BC on one side orsomething) just to make sure the code work as intended :)

**Comment by shahabgol** [comment]
> When CI passes, this seems very good to me! Only one question, would we also like a Conduction boundary condition?

Conduction BC is significant if we have two solids (with high k) in contact or the convection coefficient (h) is small, otherwise it is negligible.

File: source/solvers/heat_transfer.cc [hunk 4]
```diff
@@ -466,6 +467,24 @@ HeatTransfer<dim>::postprocess(bool first_iteration)
                       << std::endl;
         }
     }
+
+  // Minimum, maximum and average temperature
```
Comment (oguevremont) [inline source/solvers/heat_transfer.cc:471]:
```suggestion
  // Minimum and maximum temperature
```
```suggestion
  // Minimum and maximum temperature
```

File: source/solvers/heat_transfer_assemblers.cc [hunk 5]
```diff
@@ -422,18 +422,27 @@ HeatTransferAssemblerRobinBC<dim>::assemble_rhs(
...
+  const double Stefan_Boltzmann_constant =
+    this->boundary_conditions_ht.Stefan_Boltzmann_constant;
+
+  // Robin boundary condition, loop on faces (Newton's cooling law +
+  // Stefan-Boltzmann law) Convection-radiation BC is a combination of
+  // convection and radiation. If the Stefan-Boltzmann constant (with default
+  // value = 0) is set to 0, only the convection component is considered,
+  // whereas if the convection coefficient, h (with default value = 0), is set
+  // to 0, only the radiation component is considered. Otherwise, both the
+  // convection and radiation are significant on the boundary implementation
```
Comment (oguevremont) [inline source/solvers/heat_transfer_assemblers.cc:435]: Should it be explicitly said what happens when both are zero?

**Review by oguevremont — COMMENTED** [review]
Good work!

**Comment by blaisb** [comment]
@shahabgol it seems a lot of tests are breaking in debug mode?

File: source/solvers/heat_transfer_assemblers.cc (hunk 5 above)
Comment (shahabgol) [reply source/solvers/heat_transfer_assemblers.cc:435]: if both are 0, then using the convection-radiation BC is meaningless. I think it's not necessary to mention it.

**Review by oguevremont — APPROVED** [review]

**Pull request merged**